Implement Goat horns

IvanCraft623 opened the pull request:
## Introduction
Implement Goat horns and its sounds

## Follow-up
- Update [LevelSoundEvent](https://github.com/pmmp/BedrockProtocol/blob/master/src/types/LevelSoundEvent.php)

## Tests

https://user-images.githubusercontent.com/57236932/185046909-a2a8421a-8e18-484b-abea-b0af2bc6b0d3.mp4

IvanCraft623 committed:
Implement goat horns

IvanCraft623 committed:
Make PHPStan happy

IvanCraft623 committed:
...

IvanCraft623 committed:
Merge branch 'next-major' into goat-horns

IvanCraft623 committed:
Merge branch 'next-major' into goat-horns

IvanCraft623 committed:
Merge branch 'next-major' into goat-horns

IvanCraft623 committed:
Usage of LevelSoundEvent const

IvanCraft623 committed:
Merge branch 'next-major' into goat-horns

IvanCraft623 committed:
Merge branch 'next-major' into goat-horns

IvanCraft623 committed:
GoatHornTypeIdMap: fixed uninitialized fields

IvanCraft623 committed:
Merge branch 'next-major' into goat-horns

IvanCraft623 committed:
Merge branch 'minor-next' into goat-horns

dktapps commented on src/item/GoatHornType.php:
```diff
@@ -0,0 +1,83 @@
+			new self("admire", fn() => new GoatHornAdmireSound()),
+			new self("call", fn() => new GoatHornCallSound()),
+			new self("yearn", fn() => new GoatHornYearnSound()),
+			new self("dream", fn() => new GoatHornDreamSound())
```
These closures seem redundant.

IvanCraft623 replied:
Implemented a `GoatHornType::getSoundId()`, so now works similar to records

dktapps reviewed (commented):
The PR looks mostly fine, but I'm dubious about the way the sounds are handled.

I feel like the sounds could be done with much less code by having a single `GoatHornSound` class which accepts `GoatHornType` and `match()`es the sound ID. 

Is there some advantage to having a class per sound that I'm not seeing?

dktapps commented on src/item/GoatHorn.php:
```diff
@@ -0,0 +1,71 @@
+		parent::__construct($identifier, $name);
+	}
+
+	protected function encodeType(RuntimeDataWriter $w) : void{
```
This needs to be updated (encodeType no longer exists)

dktapps commented on src/item/GoatHorn.php:
```diff
@@ -0,0 +1,71 @@
+		$w->goatHornType($this->goatHornType);
+	}
+
+	public function getType() : GoatHornType{ return $this->goatHornType; }
```
Prefer more specific stuff like `getSoundType` or `getHornType` for this.

We do have `getType` in other places, but it's very generic and ambiguous, so it's best avoided.

IvanCraft623 replied:
tbh I prefer, that the code is consistent, if it is common (as observed in several places in the PM code) I would expect by intuition that to access the variants you use `getType`, so for consistency it would be strange to change the format.

dktapps replied:
My intent is to replace the getType() in other places with more specific APIs in the future. It currently only appears in 2-3 places, and the more specific style is much more common.

IvanCraft623 committed:
Merge remote-tracking branch 'upstream/minor-next' into goat-horns

IvanCraft623 committed:
Modernize code

- Usage of new native enums instead of EnumTrait
- Usage of new IntSaveIdMapTrait
- Remove unnecesary GoatHorn*Sound classes

IvanCraft623 committed:
Merge branch 'minor-next' into goat-horns

IvanCraft623 committed:
getType => getHornType

IvanCraft623 committed:
Forgot to change StringToItemParser :P

dktapps commented on src/item/GoatHornType.php:
```diff
@@ -0,0 +1,54 @@
+			self::YEARN => LevelSoundEvent::HORN_CALL6,
+			self::DREAM => LevelSoundEvent::HORN_CALL7
+		};
+	}
```
This shouldn't be exposed to the API. The only place it's used is inside GoatHornSound anyway.

IvanCraft623 replied:
Done :P

IvanCraft623 committed:
Remove GoatHornType::getSoundId() method

ShockedPlot7560 reviewed (dismissed):
PR looks fine

GamerMJay commented:
cooldown is missing :0

dktapps commented:
This looks OK, although the cooldown is indeed missing. That should be easily addressed with the merge of #6405.

ipad54 committed:
Merge branch 'minor-next' of https://github.com/pmmp/PocketMine-MP into goat-horns

ipad54 committed:
Added missing cooldown

ipad54 commented:
I've added the missing cooldown.

![Screenshot_20241111_134311](https://github.com/user-attachments/assets/92d83e9c-a641-4967-8a6f-fab0654165dc)

dktapps committed:
Merge branch 'minor-next' into goat-horns

dktapps reviewed (approved):
Looks like this is ready to go
